subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilerDecorator.java

```diff
@@ -83,4 +95,52 @@ public IncrementalCompilerDecorator(JarClasspathSnapshotMaker jarClasspathSnapsh
         PreviousCompilation previousCompilation = new PreviousCompilation(new ClassSetAnalysis(data), compileCaches.getLocalJarClasspathSnapshotStore(), compileCaches.getJarSnapshotCache());
         return new SelectiveCompiler(inputs, previousCompilation, cleaningCompiler, staleClassDetecter, compilationInitializer, jarClasspathSnapshotMaker);
     }
+
+    // Check whether there are any Annotation Processors on the classpath that are NOT
+    // incap-enabled (i.e. incremental).  If so, then we do not allow an incremental build.
+    // If there is no annotation processing at all, then we permit an incremental build.
+    private boolean allowIncrementalAnnotationProcessing() {
+        if (!compileOptions.isIncrementalAnnotationProcessing()) {
+            return false;
+        }
+        FileCollection classPath = new SimpleFileCollection(javaCompileSpec.getCompileClasspath());
+        Set<AnnotationProcessorInfo> cachedInfo = annotationProcessorDetector.getAnnotationProcessorInfo(compileOptions, classPath);
+        if (annotationProcessorDetector.checkExplicitProcessorOption(compileOptions)) {
+            LOG.info("{} is not incremental:  The explicit -processor compiler option is not supported for incremental annotation processing.", displayName);
+            return false;
+        }
+        boolean foundProcessors = false;
+        Set<String> badProcs = Sets.newHashSet();
+        for (AnnotationProcessorInfo info : cachedInfo) {
+            if (info.isProcessor()) {
```

Comment:
❌ This method shouldn't exist. All of them should be processors, otherwise there should be no ProcessorInfo.

Reply:
Actually the cache records yes/no for every artifact on the classpath.  It used to record a per-artifact boolean value in the cache, and now it records an AnnotationProcessorInfo record.  The isProcessor() flag is a way to encapsulate the old boolean.